### PR TITLE
Refactor patient genotype

### PR DESCRIPTION
**Pull request by buske**
- resolves issue #61 
- resolves issue #63
- improves test coverage and reduces checkstyle suppression for similarity-data-impl
- substantial refactor of genetics, including:
  - separated the concepts of an Exome from a PatientGenotype (which may contain exome/candidate gene data)
  - separated Abstract/Default/Restricted forms of many of the genotype classes
  - created ExomeManager component
  - created PatientGenotypeManager component
  - moved ExomiserExome handling into a specific extension of a default implementation
- removed unused getTopVariant(String, int) method of Exome
- added getTopGenes(int) method to Exome, and:
  - updated ExomiserViewScriptService to use it
  - updated remote-matching to use it: https://github.com/phenotips/remote-matching/pull/19

### CODE REVIEW

**Comment by buske** [comment]
Bump.

File: similarity-data-api/src/main/java/org/phenotips/data/similarity/ExomeManager.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package org.phenotips.data.similarity;
+
+import org.phenotips.data.Patient;
+
+import org.xwiki.stability.Unstable;
+
+/**
+ * This class allows access to an {@link #Exome} object for a given patient.
```
Comment (sdumitriu) [inline similarity-data-api/src/main/java/org/phenotips/data/similarity/ExomeManager.java]: Extra `#` in the link.

Comment (buske) [inline similarity-data-api/src/main/java/org/phenotips/data/similarity/ExomeManager.java]: I only see one hash symbol in the @link. This hash is required to get Eclipse to link up the documentation, yes?

Comment (sdumitriu) [inline similarity-data-api/src/main/java/org/phenotips/data/similarity/ExomeManager.java]: No, `#` means _member inside a class_.

1. `{@link org.package.Class}`
2. `{@link org.package.Class#field}`
3. `{@link org.package.Class#method()}`
4. `{@link ImportedClass}`
5. `{@link ImportedClass#field}`
6. `{@link ImportedClass#method()}`
7. `{@link #localField}`
8. `{@link #localMethod()}`

Comment (sdumitriu) [inline similarity-data-api/src/main/java/org/phenotips/data/similarity/ExomeManager.java]: If you can CtrlClick on that word, then it's a valid link.

File: similarity-data-api/src/main/java/org/phenotips/data/similarity/Exome.java [hunk 2]
```diff
@@ -19,23 +19,24 @@
...
 import org.xwiki.stability.Unstable;
 
+import java.util.List;
 import java.util.Set;
 
 import net.sf.json.JSONArray;
 
 /**
- * A basic representation of a genotype of a patient, with scored Variant objects in scored genes.
+ * A basic representation of a exome of a patient, with scored {@link #Variant} objects in scored genes.
```
Comment (sdumitriu) [inline similarity-data-api/src/main/java/org/phenotips/data/similarity/Exome.java]: Invalid link here.

File: similarity-data-impl/src/main/java/org/phenotips/data/similarity/script/ExomiserViewScriptService.java [hunk 3]
```diff
@@ -63,91 +60,61 @@
...
+    @Inject
+    @Named("exomiser")
+    private ExomeManager exomeManager;
+
     /**
-     * Gets the top K genes by score from a patient.
+     * Checks if a patient has a valid Exomiser genotype.
      *
-     * @param patient A valid patient
+     * @param patient A valid {@link #Patient}
```
Comment (sdumitriu) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/script/ExomiserViewScriptService.java]: The description for `@param`, `@return` and `@throws`, in case it is only one sentence, should start with lowercase and not end with a dot. It's not your fault here, though.

File: similarity-data-api/src/main/java/org/phenotips/data/similarity/ExomeManager.java (hunk 1 above)
Comment (buske) [inline similarity-data-api/src/main/java/org/phenotips/data/similarity/ExomeManager.java]: Okay, thanks for clarifying that! I'll try to use @link more responsibly in the future.

File: similarity-data-api/src/main/java/org/phenotips/data/similarity/Exome.java [hunk 4]
```diff
@@ -19,23 +19,24 @@
...
+ * @since
  */
 @Unstable
-public interface Genotype
+public interface Exome
 {
     /**
      * Get the names of all genes with variants in the patient.
      *
-     * @return an unmodifiable set of gene names with variants in the patient
```
Comment (sdumitriu) [inline similarity-data-api/src/main/java/org/phenotips/data/similarity/Exome.java]: Why did you remove `unmodifiable`?

Comment (buske) [inline similarity-data-api/src/main/java/org/phenotips/data/similarity/Exome.java]: Because when I was going through it I decided that it didn't seem necessary to have the unmodifiable constraint specified within the interface. It felt like noise?

File: similarity-data-api/src/main/java/org/phenotips/data/similarity/ExomeManager.java [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+import org.xwiki.stability.Unstable;
+
+/**
+ * This class allows access to an {@link Exome} object for a given patient.
+ *
+ * @version $Id$
+ * @since 1.0M6
+ */
+@Unstable
```
Comment (sdumitriu) [inline similarity-data-api/src/main/java/org/phenotips/data/similarity/ExomeManager.java]: `@Role`

File: similarity-data-api/src/main/java/org/phenotips/data/similarity/PatientGenotype.java [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+ * This class represents the combined genetics (candidate genes and exome sequence data) for a Patient.
+ *
+ * @version $Id$
+ * @since 1.0M6
+ */
+@Unstable
+public interface PatientGenotype extends Exome
+{
+    /**
+     * Return whether the patient has any genotype data available (candidate genes or exome data).
```
Comment (sdumitriu) [inline similarity-data-api/src/main/java/org/phenotips/data/similarity/PatientGenotype.java]: Terminology: does `candidate` mean all recorded genes, candidate or causative? Should we remove `candidate`, or replace it with something else?

File: similarity-data-api/src/main/java/org/phenotips/data/similarity/PatientGenotypeManager.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+import org.xwiki.stability.Unstable;
+
+/**
+ * This class allows access to an {@link PatientGenotype} object for a given {@link Patient}.
+ *
+ * @version $Id$
+ * @since 1.0M6
+ */
+@Unstable
```
Comment (sdumitriu) [inline similarity-data-api/src/main/java/org/phenotips/data/similarity/PatientGenotypeManager.java]: `@Role` here too.

File: similarity-data-impl/pom.xml [hunk 8]
```diff
@@ -27,7 +27,7 @@
   </parent>
 
   <properties>
-    <coverage.instructionRatio>0.15</coverage.instructionRatio>
+    <coverage.instructionRatio>0.60</coverage.instructionRatio>
```
Comment (sdumitriu) [inline similarity-data-impl/pom.xml]: I really like seeing this.

File: similarity-data-impl/src/checkstyle/checkstyle-suppressions.xml [hunk 9]
```diff
@@ -24,10 +24,4 @@
 
 <suppressions>
   <suppress checks="IllegalType" files="DefaultPatientSimilarityViewFactory.java"/>
-  <suppress checks="ClassFanOutComplexity" files="RestrictedGenotypeSimilarityView.java"/>
-  <suppress checks="ClassFanOutComplexity" files="DefaultPatientSimilarityViewFactory.java"/>
-  <suppress checks="CyclomaticComplexity" files="RestrictedGenotypeSimilarityView.java"/>
-  <suppress checks="NPathComplexity" files="RestrictedGenotypeSimilarityView.java"/>
-  <suppress checks="ClassDataAbstractionCouplingCheck" files="RestrictedGenotypeSimilarityView.java"/>
-  <suppress checks="MultipleStringLiterals" files="RestrictedGenotypeSimilarityView.java"/>
```
Comment (sdumitriu) [inline similarity-data-impl/src/checkstyle/checkstyle-suppressions.xml]: I really like seeing this too.

File: similarity-data-api/src/main/java/org/phenotips/data/similarity/PatientGenotype.java (hunk 6 above)
Comment (buske) [inline similarity-data-api/src/main/java/org/phenotips/data/similarity/PatientGenotype.java]: For now, it means whatever the Gene controller would give me, which is just what's in the candidate gene section, I believe. Prospectively, we should probably call them all `manual` genes, which have a classification level, such as `candidate`, `causative`, `previously tested`, etc.

File: similarity-data-impl/src/main/java/org/phenotips/data/similarity/script/PatientMatchingScriptService.java [hunk 10]
```diff
@@ -55,7 +59,7 @@
     public void clearCache()
     {
         ((DefaultPatientSimilarityViewFactory) patientViewFactory).clearCache();
-        PatientGenotype.clearCache();
+        ((ExomiserExomeManager) exomeManager).clearCache();
```
Comment (sdumitriu) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/script/PatientMatchingScriptService.java]: That's not nice, don't make assumptions about the implementation. Maybe add the `clearCache` method in `ExomeManager`?

Comment (buske) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/script/PatientMatchingScriptService.java]: I considered that, but figured it was too implementation-specific to be included in the interface. Same with the `PatientSimilarityViewFactory`. :/

File: similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultGenotypeSimilarityView.java [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        try {
+            return componentManager.getInstance(PatientGenotypeManager.class);
+        } catch (ComponentLookupException e) {
+            logger.error("Unable to look up PatientGenotypeManager: " + e.toString());
+        }
+        return null;
+    }
+
+    /**
+     * Score genes with variants in both patients, and sets 'geneScores' and 'geneVariants' fields accordingly.
```
Comment (sdumitriu) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultGenotypeSimilarityView.java]: The comment and the code are contradicting on this point: `genes with variants in both patients`. Which one is correct?

File: similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/AbstractExome.java [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+    @Override
+    public Set<String> getGenes()
+    {
+        return Collections.unmodifiableSet(this.variants.keySet());
+    }
+
+    @Override
+    public Double getGeneScore(String gene)
+    {
+        return this.geneScores.get(gene);
```
Comment (sdumitriu) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/AbstractExome.java]: Possible NPE, `geneScores` may be null.

File: similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultGenotypeSimilarityView.java [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+     */
+    private void matchGenes()
+    {
+        this.geneScores = new HashMap<String, Double>();
```
Comment (sdumitriu) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultGenotypeSimilarityView.java]: `geneScores` was just initialized in the constructor.

File: similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultGenotypeSimilarityView.java (hunk 11 above)
Comment (buske) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultGenotypeSimilarityView.java]: Hmm, not sure I quite see the contradiction, but perhaps "genes found in both patients, and sets 'geneScores' field accordingly"?

Comment (sdumitriu) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultGenotypeSimilarityView.java]: The contradiction is that the code doesn't care if a gene is in both sources, it will just use `0` as the score for the missing one. Or does `getGenes` do the filtering on pairs?

File: similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/AbstractExome.java (hunk 12 above)
Comment (buske) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/AbstractExome.java]: Good spot. Is the appropriate way of protecting against this to initialize `geneScores` in a constructor in the abstract class, or just check for null in the method (or both)?

File: similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultGenotypeSimilarityView.java (hunk 11 above)
Comment (sdumitriu) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultGenotypeSimilarityView.java]: Yes it does...

Comment (buske) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultGenotypeSimilarityView.java]: Yeah, for the PatientGenotypeSimilarityView, `getGenes` is currently those in the intersection.

File: similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/AbstractExome.java (hunk 12 above)
Comment (sdumitriu) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/AbstractExome.java]: Safest is to always check for null, but it should be enough to initialize it when declaring the field.

File: similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultPatientSimilarityView.java [hunk 14]
```diff
@@ -368,7 +369,7 @@ public double getScore()
      *
      * @return the genotype similarity view for this pair of patients
      */
-    private GenotypeSimilarityView getGenotypeSimilarity()
+    private PatientGenotypeSimilarityView getGenotypeSimilarity()
     {
         if (this.matchedGenes == null) {
             this.matchedGenes = new RestrictedGenotypeSimilarityView(this.match, this.reference, this.access);
```
Comment (sdumitriu) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultPatientSimilarityView.java]: I wonder why does the `Default` patient view return a `Restricted` genotype view...

Comment (sdumitriu) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultPatientSimilarityView.java]: It was like that from the start: https://github.com/phenotips/patient-network/commit/0e2e18846074895f82103c5411714e25741bd5b7#diff-c9706f062e531bc7b3f78e4cad9ca7d6R334

Comment (buske) [inline similarity-data-impl/src/main/java/org/phenotips/data/similarity/internal/DefaultPatientSimilarityView.java]: Indeed, it shouldn't anymore. Initially there was *only* the `Restricted`
genotype view. Now that there is a `Default`, it should return that insted.
